File: packages/native/runtime-swift/Sources/PyreonRuntime/PyreonCrashReporter.swift
// PyreonCrashReporter — the native runtime container `useCrashReporter()`
// lowers to. The CAPTURE + PERSIST + REHYDRATE half of crash reporting,
// credential-free by design: the vendor TRANSPORT (Sentry, Crashlytics, a
// custom endpoint) is app-wired through `PyreonCrashTransportRegistry`, so
// the framework proves the part no credential gates — an uncaught error is
// captured, written to disk, and readable on the NEXT launch — and an app
// adds shipping-grade upload with one closure.
//
//     let crash = PyreonCrashReporter()
//     crash.start()          // installs the NSException hook + rehydrates
//     crash.lastCrash        // previous launch's report, "" when none
//     crash.recordError("…") // manual capture (caught errors, assertions)
//     crash.breadcrumb("…")  // ring buffer, attached to the next report
//
// HONEST SCOPE (v1): `NSSetUncaughtExceptionHandler` captures uncaught
// NSExceptions; SIGNAL crashes (SIGSEGV/SIGABRT from Swift fatalError and
// memory errors) are NOT captured — signal-safe crash handling is a
// substantial follow-up (what vendor SDKs exist for). `recordError` covers
// the manual path today. Stated here and in the matrix rather than implied
// away.
//
// The capture STATE (record/breadcrumb/rehydrate/clear) is pure and
// unit-testable against a temp directory; only the exception-hook wiring in
// `start()` needs a real process to prove. Mirrors PyreonAppState's
// start()-idempotency + sticky-assertable-state design.

import Foundation
import Observation

/// App-wired transport seam — same shape as the push/geolocation registries:
/// a framework default that would need a vendor credential does not exist,
/// so the slot is explicit and the framework never fakes an upload.
public enum PyreonCrashTransportRegistry {
    /// Called with the previous launch's report on `start()` rehydrate, and
    /// with each `recordError` report. Wire in Application setup.
    public static var send: ((String) -> Void)?
}

@available(iOS 17.0, macOS 14.0, *)
@Observable
public final class PyreonCrashReporter {
    /// The PREVIOUS launch's crash report (JSON string), "" when the last
    /// launch exited cleanly. Populated by `start()` from the persisted
    /// file — the rehydrate half of the device proof.
    public private(set) var lastCrash: String = ""

    /// STICKY: true once THIS launch found a persisted report. The
    /// device-test assertion surface (the PyreonAppState `wasBackgrounded`
    /// pattern): an end-state a never-started reporter can never reach.
    public private(set) var hadCrash: Bool = false

    @ObservationIgnored private var _started = false
    @ObservationIgnored private var breadcrumbs: [String] = []
    @ObservationIgnored private let dir: URL

    /// Whether the exception hook is installed. Not observable — a status
    /// read must not force a re-render.
    public var isMonitoring: Bool { _started }

    /// `directory` injectable for tests; defaults to Application Support,
    /// which survives relaunches (the whole point).
    public init(directory: URL? = nil) {
        if let directory {
            dir = directory
        } else {
            let base = FileManager.default.urls(
                for: .applicationSupportDirectory, in: .userDomainMask
            ).first ?? FileManager.default.temporaryDirectory
            dir = base.appendingPathComponent("pyreon-crash", isDirectory: true)
        }
    }

    private var reportURL: URL { dir.appendingPathComponent("last-crash.json") }

    /// Install the uncaught-NSException hook + rehydrate any persisted
    /// report. Idempotent. The hook writes SYNCHRONOUSLY — the process is
    /// about to die, there is no later.
    public func start() {
        guard !_started else { return }
        _started = true
        rehydrate()
        // The C-function hook cannot capture `self`; route through a static
        // slot (one live reporter owns crash capture — the deep-link
        // single-slot doctrine: an append-only list on a global is the
        // unbounded-growth shape).
        PyreonCrashReporter.active = self
        NSSetUncaughtExceptionHandler { exception in
            PyreonCrashReporter.active?.persist(
                message: "\(exception.name.rawValue): \(exception.reason ?? "")",
                stack: exception.callStackSymbols.joined(separator: "\n")
            )
        }
    }

    @ObservationIgnored nonisolated(unsafe) static var active: PyreonCrashReporter?

    /// Manual capture — caught errors, assertion failures, "should never
    /// happen" branches. Persists AND forwards to the transport when wired.
    public func recordError(_ message: String) {
        persist(message: message, stack: Thread.callStackSymbols.joined(separator: "\n"))
        PyreonCrashTransportRegistry.send?(message)
    }

    /// Ring-buffered context attached to the next report (capped — a crash
    /// reporter must never be its own unbounded-growth bug).
    public func breadcrumb(_ message: String) {
        breadcrumbs.append(message)
        if breadcrumbs.count > 32 { breadcrumbs.removeFirst() }
    }

    /// Acknowledge the rehydrated report (e.g. after showing an apology UI
    /// or forwarding it). Clears the observable state AND the file.
    public func clear() {
        lastCrash = ""
        hadCrash = false
        try? FileManager.default.removeItem(at: reportURL)
    }

    // MARK: - internals (pure vs the injected directory — unit-tested)

    func persist(message: String, stack: String) {
        let report: [String: Any] = [
            "message": message,
            "stack": stack,
            "breadcrumbs": breadcrumbs,
            "timestamp": ISO8601DateFormatter().string(from: Date()),
        ]
        guard let data = try? JSONSerialization.data(withJSONObject: report) else { return }
        try? FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        try? data.write(to: reportURL, options: .atomic)
    }

    func rehydrate() {
        guard let data = try? Data(contentsOf: reportURL),
              let text = String(data: data, encoding: .utf8), !text.isEmpty
        else { return }
        lastCrash = text
        hadCrash = true
        PyreonCrashTransportRegistry.send?(text)
        // Deliberately NOT cleared on read: the app decides when the report
        // is handled (`clear()`), so a transport wired one launch late still
        // sees it.
    }
}

File: packages/fundamentals/hooks/native/kotlin/com/pyreon/runtime/PyreonCrashReporter.kt
// PyreonCrashReporter — the native runtime container `useCrashReporter()`
// lowers to. The CAPTURE + PERSIST + REHYDRATE half of crash reporting,
// credential-free by design: the vendor TRANSPORT (Sentry, Crashlytics, a
// custom endpoint) is app-wired through `PyreonCrashTransportRegistry`, so
// the framework proves the part no credential gates — an uncaught error is
// captured, written to disk, and readable on the NEXT launch.
//
//     val crash = PyreonCrashReporter(backend)
//     crash.start()                 // installs the default handler + rehydrates
//     crash.lastCrash.value         // previous launch's report, "" when none
//     crash.recordError("…")        // manual capture (caught errors)
//     crash.breadcrumb("…")         // ring buffer, attached to the next report
//
// The uncaught handler CHAINS to the previous default handler — the process
// still dies and the system crash dialog still shows; a crash reporter that
// swallows the crash changes app behavior, which capture must never do.
//
// **Persistence is backend-injected** (the `PyreonStorageBackend` shape the
// database/storage files use) so this file stays free of `android.*`/`java.io`
// beyond what the kotlinc validate stubs carry, the capture state is fully
// unit-testable against an in-memory backend, and a cold INSTANCE over the
// same backend models a relaunch — the same documented in-process ceiling as
// the Android auth-rehydration proof. `rememberPyreonCrashReporter()` (the
// Android-side file) self-installs a file-backed default: a default that
// requires a step nobody takes is not a default.
//
// HONEST SCOPE (v1): `Thread.setDefaultUncaughtExceptionHandler` captures
// JVM uncaught throwables; NATIVE (NDK) crashes are not captured — the same
// disclosed boundary as the Swift half's signal crashes.

package com.pyreon.runtime

import androidx.compose.runtime.MutableState
import androidx.compose.runtime.mutableStateOf

/** App-wired transport seam — the push/geolocation registry shape. */
public object PyreonCrashTransportRegistry {
    /** Called with the rehydrated report on start(), and with each recordError. */
    public var send: ((String) -> Unit)? = null
}

public class PyreonCrashReporter(
    private val backend: PyreonStorageBackend = InMemoryBackend(),
) {
    /** Previous launch's crash report (JSON string); "" when none. */
    public val lastCrash: MutableState<String> = mutableStateOf("")

    /**
     * STICKY: true once THIS launch found a persisted report — the
     * device/instrumented assertion surface (the `wasBackgrounded` pattern).
     */
    public val hadCrash: MutableState<Boolean> = mutableStateOf(false)

    private var started = false
    private val breadcrumbs = ArrayDeque<String>()

    /** Whether the uncaught handler is installed. */
    public val isMonitoring: Boolean get() = started

    /**
     * Install the uncaught-throwable handler + rehydrate any persisted
     * report. Idempotent. The handler persists SYNCHRONOUSLY, then CHAINS to
     * the previous default handler so the process still dies normally.
     */
    public fun start() {
        if (started) return
        started = true
        rehydrate()
        val previous = Thread.getDefaultUncaughtExceptionHandler()
        Thread.setDefaultUncaughtExceptionHandler { thread, error ->
            persist(
                message = "${error::class.simpleName}: ${error.message ?: ""}",
                stack = error.stackTraceToString(),
            )
            previous?.uncaughtException(thread, error)
        }
    }

    /** Manual capture — persists AND forwards to the transport when wired. */
    public fun recordError(message: String) {
        persist(message = message, stack = Throwable().stackTraceToString())
        PyreonCrashTransportRegistry.send?.invoke(message)
    }

    /** Ring-buffered context attached to the next report (capped at 32). */
    public fun breadcrumb(message: String) {
        breadcrumbs.addLast(message)
        while (breadcrumbs.size > 32) breadcrumbs.removeFirst()
    }

    /** Acknowledge the rehydrated report: clears state AND the persisted key. */
    public fun clear() {
        lastCrash.value = ""
        hadCrash.value = false
        backend.remove(KEY)
    }

    // internals — pure vs the injected backend (unit-tested)

    internal fun persist(message: String, stack: String) {
        // PyreonJson.encode gives correct JSON string escaping - a hand-built
        // JSON string would re-create the delimiter-escaping class for stack
        // traces containing quotes/newlines.
        val crumbs = breadcrumbs.joinToString("\n")
        backend.write(KEY, PyreonJson.encode(listOf(message, stack, crumbs)))
    }

    private fun rehydrate() {
        val raw = backend.read(KEY) ?: return
        if (raw.isEmpty()) return
        lastCrash.value = raw
        hadCrash.value = true
        PyreonCrashTransportRegistry.send?.invoke(raw)
        // Deliberately NOT cleared on read — the app decides via clear(),
        // so a transport wired one launch late still sees the report.
    }

    private companion object {
        const val KEY: String = "pyreon.crash.last"
    }
}
